src/Bundle.ts

```diff
@@ -12,68 +12,349 @@ import transformBundle from './utils/transformBundle';
+	name: string;
 	graph: Graph;
 	orderedModules: Module[];
+
+	// this represents the bundle module wrappings
+	// which form the output dependency graph
+
+	// map from variable exported by this bundle to its safe exported name
+	exportedVariables: Map<Variable, string>;
+	imports: {
+		module: Bundle | ExternalModule;
+		variables: {
+			// the name of the export this import corresponds to
+			name: string;
+			module: Module | ExternalModule;
+			variable: Variable;
+		}[];
+	}[];
+	exports: {
+		[safeName: string]: {
+			// module can be in or out of bundle
+			// if module is out of the bundle then it is a reexport
+			module: Module | ExternalModule;
+			// exported name from source module
+			name: string;
+			variable: Variable;
+		}
+	}
+	dependencies: (ExternalModule | Bundle)[];
 	externalModules: ExternalModule[];
+	// an entry module bundle is a bundle that exactly exports the exports of
+	// an input entry point module
 	entryModule: Module;
+	entryModuleFacade: boolean;
 
-	constructor (graph: Graph, orderedModules: Module[]) {
+	constructor (graph: Graph, id: string, orderedModules: Module[]) {
+		this.setId(id);
 		this.graph = graph;
 		this.orderedModules = orderedModules;
+
+		this.exportedVariables = new Map();
+		this.imports = [];
+		this.exports = {};
 		this.externalModules = undefined;
+
+		this.dependencies = undefined;
 		this.entryModule = undefined;
+		this.entryModuleFacade = false;
+		orderedModules.forEach(module => {
+			if (module.isEntryPoint) {
+				if (!this.entryModule) {
+					this.entryModule = module;
+					this.entryModuleFacade = true;
+				} else {
+					this.entryModuleFacade = false;
+				}
+			}
+			module.bundle = this;
+		});
 	}
 
-	bind () {
-		this.orderedModules.forEach(module => module.bindImportSpecifiers());
-		this.orderedModules.forEach(module => module.bindReferences());
+	setId (id: string) {
+		this.id = id;
+		this.name = makeLegal(id);
 	}
 
-	includeMarked (treeshake: boolean) {
-		if (treeshake) {
-			let addedNewNodes;
-			do {
-				addedNewNodes = false;
-				this.orderedModules.forEach(module => {
-					if (module.includeInBundle()) {
-						addedNewNodes = true;
-					}
-				});
-			} while (addedNewNodes);
+	// ensure that the module exports or reexports the given variable
+	// we don't replace reexports with the direct reexport from the final module
+	// as this might result in exposing an internal module which taints an entryModule bundle
+	ensureExport (module: Module | ExternalModule, variable: Variable): string {
+		let safeExportName = this.exportedVariables.get(variable);
+		if (safeExportName) {
+			return safeExportName;
+		}
+
+		let i = 0;
+		if (variable.exportName) {
+			safeExportName = variable.exportName;
 		} else {
-			// Necessary to properly replace namespace imports
-			this.orderedModules.forEach(module => module.includeAllInBundle());
+			safeExportName = variable.exportName = variable.name;
+		}
+
+		let curExport = this.exports[safeExportName];
```

Comment:
`curExport` has a double meaning: First it is used to iterate through existing exports, then it is used to contain the new export. I would suggest the following change:

```ts
while (this.exports[safeExportName]) safeExportName = (variable.exportName || variable.name) + '$' + ++i

const newExport = this.exports[safeExportName] = { module, name: undefined, variable }
```
and then use `newExport` instead of `curExport` from then on.
Also, this produces names like `myShadowedExport$1$2$3` (if I am not mistaken). Would be cooler (though just cosmetically) to have names like `myShadowedExport$3` instead. As conflicting exports are probably not very common, I do not think this is too performance critical as long as the unshadowed version is fast.

Reply:
Sure. This shouldn't produce names like that as both the `exportName` and `name` are the original names due to the rendering changes.

Reply:
Ah you are right, misread this.